doc/ACL_pytest.md

```diff
@@ -0,0 +1,326 @@
+
+NOTE: *We will skip this part here in the design right now until we reach some decision on this limitation; Below fixutres in this design depend on 'dut' host provided by 'ansible_adhoc'*
+
+
+##### Common tasks missing
+
+The following has to be implemented in the same way as in relevant playbook roles:
+
+- reboot
+- reload config
+- port toggle
+
+### Design
+
+#### Directory structure
+
+The test will be structured as follows:
+
+```
+tests/
+    acstests/ -> ../ansible/roles/test/files/acstests/
+    common/
+        __init__.py
+        reload_config.py
+        reboot.py
+        port_toggle.py
+    acl/
+        conftest.py
+        test_acl.py
+        templates/
+        loganalyzer/
+```
+
+- The *acl* directory here created, so that other test modules related to ACL test can be put here and ACL related fixtures, etc. can live in conftest.py
+- The *common* has common functionallity required by a set of tests.
+- *acstests* simply a symbolic link to acs PTF test scripts
+
+
+#### test_acl.py
+
+The current test is divided into 4 cases:
+- Basic test - setup ACL rules, run traffic, teardown ACL rules
+- Incremental test - setup ACL rules incrementaly, run traffic, teardown ACL rules
+- Reboot test - setup ACL rules, reboot, run traffic, teardown ACL rules
+- Port toggle test - setup ACL rules, toggle all ports, run traffic, teardown ACL rules
+
+The test logic will be seperated into generic traffic test in *BaseAclTest* and derivatives are meant to provide fixture to setup/teardown according to their requirements.
+
+According to above cases 4 test classes has to be coded:
+
+- *TestAcl*
+- *TestIncrementalAcl*
+- *TestAclWithReboot*
+- *TestAclWithPortToggle*
+
+
+In common for all above cases we need to do the following setup/teardown:
+- setup test variables according to topology and minigraph facts
+e.g. get the list of ToR/Spine ports, setup names and configuration for ACL tables and so on
+- generate and deploy configuration files to the DUT, after the test remove configuration and cleanup
+
+Also configuration generation and dependent traffic run tests will be parametrized for 'ingress', 'egress' ACL tests
+
+##### Test fixtures
+
+As above described setup and configuration generation can be done in scope of *test_acl.py* module only once.
+
+These fixtures, by the way,can live in *acl/conftest.py*, so other ACL test modules can reuse them.
+
+###### Module scope
+
+```python
+@pytest.fixture(scope='module')
+def setup(dut, testbed):
+    '''
+    Setup fixture gathers all test required information from DUT facts and testbed
+    '''
+
+@pytest.fixture(scope='module', params=['ingress', 'egress'])
+def stage(request):
+    '''
+    Small fixture to return the ACL table stage is beeing tested
+    '''
+
+@pytest.yield_fixture(scope='module')
```

Comment:
According to this announcement https://docs.pytest.org/en/latest/yieldfixture.html, the `yield_fixture` function is deprecated and should not be used in new code. Since pytest-3.0, fixtures can directly use the `yield` statement.

Reply:
Thanks, in the code I will consider this for sure. For this design document I think this is more clear that fixture has code before and after tests